### PR TITLE
graph-tool: import from homebrew/science

### PR DESCRIPTION
**Pull request by songololo**
matplotlib in venv not yet resolved (how to link?)

NOTE: 
- Matplotlib requires different resources based on python2 vs python3
- Installing resources manually because `venv.pip_install_and_link buildpath` fails because it can't find a `setup.py`.
- Using `venv.pip_install_and_link resource(r)` method but this does not successfully link `matplotlib` to the brewed application?

COMMENTS:
- The brewed package does not find a linked `matplotlib`
- The brewed package works if `matplotlib` is installed separately via `pip3 install matplotlib` (though the point is for it to work without this step).
- FYI: The `graph-tool` build takes quite a long time to complete.

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

brew audit throws these errors (not sure how to resolve)
```
  * C: 26: col 3: :python is deprecated. Usage should be "python".
  * C: 27: col 3: :python3 is deprecated. Usage should be "python3".
  * C: 42: col 16: Dependency librsvg should not use option with-gtk+3
```
-----

### CODE REVIEW

**Comment by shongololo** [comment]
FYI: If importing into python3 (without a separate install of `matplotlib`:
```
import graph_tool.all as gt
/usr/local/lib/python3.6/site-packages/graph_tool/draw/cairo_draw.py:67: RuntimeWarning: Error importing matplotlib module. Graph drawing will not work.
  warnings.warn(msg, RuntimeWarning)
/usr/local/lib/python3.6/site-packages/graph_tool/draw/cairo_draw.py:67: RuntimeWarning: Error importing matplotlib module. Graph drawing will not work.
  warnings.warn(msg, RuntimeWarning)
/usr/local/lib/python3.6/site-packages/graph_tool/all.py:40: RuntimeWarning: Error importing draw module, proceeding nevertheless: No module named 'matplotlib'
  warnings.warn(msg, RuntimeWarning)
```
Works once `matplotlib` is installed via `pip3` (however the point is for this not to be necessary.

Example of how to test in python3:
```
import graph_tool.all as gt
g = gt.Graph()
v1 = g.add_vertex()
v2 = g.add_vertex()
e = g.add_edge(v1, v2)
gt.graph_draw(g)
```
Which should plot a graph, assuming matplotlib is found.

**Comment by ilovezfs** [comment]
```
diff --git a/Formula/graph-tool.rb b/Formula/graph-tool.rb
index 5af6d5dabb..2a75be0b6f 100644
--- a/Formula/graph-tool.rb
+++ b/Formula/graph-tool.rb
@@ -132,6 +132,9 @@ class GraphTool < Formula
           system "make", "install"
         end
       end
+      site_packages = "lib/python#{version}/site-packages"
+      pth_contents = "import site; site.addsitedir('#{libexec/site_packages}')\n"
+      (prefix/site_packages/"homebrew-graph-tool.pth").write pth_contents
     end
   end
 
```
I also rebuilt `boost-python` with `--with-python3` and `pygobject3` with `--with-python3`.

screen shot 2018-01-13 at 8 45 41 am

**Comment by ilovezfs** [comment]
That said, the formula probably cannot be added with `python3` support if it requires

```
depends_on "foo" => "with-bar"
```
which is an audit violation, which it seems to need at least for `boost-python` and `pygobject3`.

Also, it probably cannot be added with `gtk+3` support either due to the
```
depends_on "librsvg" => "with-gtk+3"
```

**Comment by shongololo** [comment]
Thanks for showing how to link the path. I'm rebuilding and testing and will report back if anything odd happens.

Regarding the `=> with-python3` type syntax, is there a way around this? Seems odd not to be able to install packages with python3 support?

Thanks.

**Comment by ilovezfs** [comment]
>is there a way around this?

The relevant formulae would need to be changed to build with both python2 and python3 support simultaneously, or new, and separate, `foo-python3` formulae would need to be added to homebrew/core, and then you could depend on both e.g. `boost-python` and `boost-python3`.

**Comment by shongololo** [comment]
Would it be OK if I make pull requests to change those formulas to include python3? e.g.
`depends_on "python3" => :recommended`

**Comment by ilovezfs** [comment]
@tschoonj any thoughts on `pygobject3` having both python and python3 be `:recommended`?

**Comment by tschoonj** [comment]
We had this discussion before. IMHO `pygobject3` should be split into 2 formulas `py2gobject3` and `py3gobject3`, similar to `py2cairo` and `py3cairo`. I opened a PR for this a couple of months ago but we couldn't agree on names for the formulas so nothing happened...

**Comment by ilovezfs** [comment]
@shongololo so that sounds like a "no" to me.

**Comment by shongololo** [comment]
@ilovezfs I tried reinstalling graph-tool with the above modifications but it still fails to find / link to matplotlib (unless I explicitly install matplotlib vis pip3).

**Comment by ilovezfs** [comment]
python 2 or 3?

**Comment by shongololo** [comment]
I am trying for python3

**Comment by ilovezfs** [comment]
what's the output of `cat /usr/local/lib/python3.6/site-packages/homebrew-graph-tool.pth`

**Comment by shongololo** [comment]
@ilovezfs this is what I get:
`import site; site.addsitedir('/usr/local/Cellar/graph-tool/2.26/libexec/lib/python3.6/site-packages')`

**Comment by ilovezfs** [comment]
and what's the output of
```
/usr/local/bin/python3 -c 'import matplotlib; print(matplotlib.__version__)'
```

**Comment by shongololo** [comment]
@ilovezfs 
```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'matplotlib'
```
Though I've intentionally removed matplotlib (via pip3) to check that the installation in venv via graph-tool works. (If matplotlib is already installed, or if it is installed separately via pip3, then graph-tool works fine as it finds the existing installation.)

**Comment by shongololo** [comment]
@ilovezfs @count0 I'm wondering if we can get this discussion going again to see if there is a way to break the logjam on getting graph-tool migrated into core?

As I understand it the problems are:
1. Cannot use `depends_on` in formulas, i.e. these lines somehow have to be resolved:
  - `depends_on "boost-python" => "with-python3"`
  - `depends_on "librsvg" => "with-gtk+3"`
  - `depends_on "pygobject3" => "with-python3"`

Am I right that the only way to resolve the `with-python3` lines is to convince the maintainers of the required formulas to create dedicated `python3` variants that can be installed in either python2 or python3 variants depending?

Is there anyway to resolve the `librsvg` line?

2.  We somehow need to figure out how to install and properly link `matplotlib` per the ongoing discussion.

**Comment by shongololo** [comment]
A heads-up that this is almost working except I can't find a way around this line:
`depends_on "pygobject3" => with-python3`

This will be resolved once the `pygobject3` naming issue is resolved for python2 vs python3 per Homebrew/homebrew-core#18711, but in the interim I don't know what to suggest.

**Comment by shongololo** [comment]
I've updated the pull request with a simplified formula intended for python 3 only.

For now I'm using `depends_on "pygobject3" => "with-python3"` as a workaround until the future reworking / renaming for python3 is completed.

**Comment by shongololo** [comment]
Updating to mark as not stale.

Looking through the current `pygobject3` script it appears that `python3` is now the default, in which case all issues should be resolved.

**Comment by kikohs** [comment]
Any updates on this?

**Comment by shongololo** [comment]
@ilovezfs would it be possible to get this merged? I've been using the graph-tool brew script on my system with no issues. Should work with the latest python3 changes etc., but please let me know if there is anything else.

**Comment by syassami** [comment]
Works on my system as well, (High Sierra)

**Comment by shongololo** [comment]
@MikeMcQuaid would it be possible to have this merged? It appears that all is working as intended and `python3` vs. `pygobject3` related issues are now resolved. Thanks.

**Comment by ilovezfs** [comment]
There's an audit failure about using explicit framework links.
```
==> brew audit graph-tool --online --new-formula
==> FAILED
Error: 2 problems in 1 formula
graph-tool:
  * C: 17: col 3: dependency "librsvg" (line 17) should be put before dependency "macos" (line 16)
  * python modules have explicit framework links
    These python extension modules were linked directly to a Python
    framework binary. They should be linked with -undefined dynamic_lookup
    instead of -lpython or -framework Python.
      /usr/local/opt/graph-tool/lib/python3.6/site-packages/graph_tool/centrality/libgraph_tool_centrality.0.so
            /usr/local/opt/graph-tool/lib/python3.6/site-packages/graph_tool/centrality/libgraph_tool_centrality.so
            /usr/local/opt/graph-tool/lib/python3.6/site-packages/graph_tool/clustering/libgraph_tool_clustering.0.so
            /usr/local/opt/graph-tool/lib/python3.6/site-packages/graph_tool/clustering/libgraph_tool_clustering.so
            /usr/local/opt/graph-tool/lib/python3.6/site-packages/graph_tool/correlations/libgraph_tool_correlations.0.so
            /usr/local/opt/graph-tool/lib/python3.6/site-packages/graph_tool/correlations/libgraph_tool_correlations.so
            /usr/local/opt/graph-tool/lib/python3.6/site-packages/graph_tool/draw/libgraph_tool_draw.0.so
            /usr/local/opt/graph-tool/lib/python3.6/site-packages/graph_tool/draw/libgraph_tool_draw.so
            /usr/local/opt/graph-tool/lib/python3.6/site-packages/graph_tool/draw/libgraph_tool_layout.0.so
            /usr/local/opt/graph-tool/lib/python3.6/site-packages/graph_tool/draw/libgraph_tool_layout.so
            /usr/local/opt/graph-tool/lib/python3.6/site-packages/graph_tool/flow/libgraph_tool_flow.0.so
            /usr/local/opt/graph-tool/lib/python3.6/site-packages/graph_tool/flow/libgraph_tool_flow.so
            /usr/local/opt/graph-tool/lib/python3.6/site-packages/graph_tool/generation/libgraph_tool_generation.0.so
            /usr/local/opt/graph-tool/lib/python3.6/site-packages/graph_tool/generation/libgraph_tool_generation.so
            /usr/local/opt/graph-tool/lib/python3.6/site-packages/graph_tool/inference/libgraph_tool_inference.0.so
            /usr/local/opt/graph-tool/lib/python3.6/site-packages/graph_tool/inference/libgraph_tool_inference.so
            /usr/local/opt/graph-tool/lib/python3.6/site-packages/graph_tool/libgraph_tool_core.0.so
            /usr/local/opt/graph-tool/lib/python3.6/site-packages/graph_tool/libgraph_tool_core.so
            /usr/local/opt/graph-tool/lib/python3.6/site-packages/graph_tool/search/libgraph_tool_search.0.so
            /usr/local/opt/graph-tool/lib/python3.6/site-packages/graph_tool/search/libgraph_tool_search.so
            /usr/local/opt/graph-tool/lib/python3.6/site-packages/graph_tool/spectral/libgraph_tool_spectral.0.so
            /usr/local/opt/graph-tool/lib/python3.6/site-packages/graph_tool/spectral/libgraph_tool_spectral.so
            /usr/local/opt/graph-tool/lib/python3.6/site-packages/graph_tool/stats/libgraph_tool_stats.0.so
            /usr/local/opt/graph-tool/lib/python3.6/site-packages/graph_tool/stats/libgraph_tool_stats.so
            /usr/local/opt/graph-tool/lib/python3.6/site-packages/graph_tool/topology/libgraph_tool_topology.0.so
            /usr/local/opt/graph-tool/lib/python3.6/site-packages/graph_tool/topology/libgraph_tool_topology.so
            /usr/local/opt/graph-tool/lib/python3.6/site-packages/graph_tool/util/libgraph_tool_util.0.so
            /usr/local/opt/graph-tool/lib/python3.6/site-packages/graph_tool/util/libgraph_tool_util.so
```

**Comment by shongololo** [comment]
Would it be OK to add back the undefined dynamic lookup config args?
```
"PYTHON_LDFLAGS=-undefined dynamic_lookup",
"PYTHON_LIBS=-undefined dynamic_lookup",
"PYTHON_EXTRA_LIBS=-undefined dynamic_lookup"
```

**Comment by ilovezfs** [comment]
It would be good to determine which one is actually needed. Probably just `PYTHON_LDFLAGS`

File: Formula/graph-tool.rb [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+  resource "six" do
+    url "https://files.pythonhosted.org/packages/16/d8/bc6316cf98419719bd59c91742194c111b6f2e85abac88e496adefaf7afe/six-1.11.0.tar.gz"
+    sha256 "70e8a77beed4562e7f14fe23a786b54f6296e34344c23bc42f07b15018ff98e9"
+  end
+
+  def install
+    xy = Language::Python.major_minor_version "python3"
+
+    inreplace "configure", "libboost_python", "libboost_python36"
```
Comment (ilovezfs) [inline Formula/graph-tool.rb]: Is there some way we can pass this and the line below to configure? Upstream needs to be able to handle the Boost 1.67.0 Python libraries.

File: Formula/graph-tool.rb [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+                          "--with-python-module-path=#{lib}/python#{xy}/site-packages"
+    system "make", "install"
+
+    site_packages = "lib/python#{xy}/site-packages"
+    pth_contents = "import site; site.addsitedir('#{libexec/site_packages}')\n"
+    (prefix/site_packages/"homebrew-graph-tool.pth").write pth_contents
+  end
+
+  test do
+    Pathname("test.py").write <<~EOS
```
Comment (ilovezfs) [inline Formula/graph-tool.rb]:
```
(testpath/"test.py").write <<~EOS
```

**Comment by shongololo** [comment]
@count0 are you able to comment regarding the question (two comments above) about passing arguments to configure?

**Comment by ilovezfs** [comment]
(Just for context, there was a change in the latest Boost that causes the Boost Python libraries to get "27" or "36" always appended to the library name for Python 2.7 and Python 3.6, respectively, which Boost-using projects need to adapt to.)

**Comment by count0** [comment]
@shongololo This can be done with the configure option:
```
--with-boost-python     specify yes/no or the boost python library or suffix to use
```
So, you only need to do `--with-boost-python=36`, or `--with-boost-python=libboost-python36`.

**Comment by ilovezfs** [comment]
Thanks for your first contribution to Homebrew, @shongololo! Without people like you submitting PRs we couldn't run this project. You rock!

**Comment by shongololo** [comment]
@ilovezfs thanks for your patience and helping us get this one figured out and off the ground!

@count0 with the next brew update it should now be possible to install via brew on mac OS (python3 only)

**Comment by ilovezfs** [comment]
You're welcome!

**Comment by count0** [comment]
@shongololo thank you very much for your effort! I'll update the website documentation accordingly.

**Pull request closed without merging**